sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/options/S3Options.java

```diff
@@ -33,13 +37,44 @@
 
   @Description(
       "Size of S3 upload chunks; max upload object size is this value multiplied by 10000;"
-          + "default is 64MB, or 5MB in memory-constrained environments")
-  @Nullable
+          + "default is 64MB, or 5MB in memory-constrained environments. Must be at least 5MB.")
+  @Default.InstanceFactory(S3UploadBufferSizeBytesFactory.class)
   Integer getS3UploadBufferSizeBytes();
+
   void setS3UploadBufferSizeBytes(Integer value);
 
   @Description("Thread pool size, limiting max concurrent S3 operations")
   @Default.Integer(50)
   int getS3ThreadPoolSize();
   void setS3ThreadPoolSize(int value);
+
+  @Description("Algorithm for SSE-S3 encryption, e.g. AES256.")
+  @Nullable
+  String getSSEAlgorithm();
+  void setSSEAlgorithm(String value);
+
+  @Description("SSE key for SSE-C encryption, e.g. a base64 encoded key and the algorithm.")
+  @Nullable
+  SSECustomerKey getSSECustomerKey();
```

Comment:
Since this is a custom object type, how does it get serialized/deserialized with Jackson?

Reply:
Hi, thanks you have a really valid argument. I hesitated to put the arguments as Strings and build the API objects internally but I thought letting the AWS SDK exposed would move this maintenance to the user. Do you have any suggestion on how to do this properly ?

Reply:
I would suggest following the same pattern that was done for the AWSCredentialsProvider. You can see how the class is being converted to and from JSON within AwsModule.java. You effectively define the schema of the object as JSON.

AWSCredentialsProvider to/from JSON had some complexity since it had to deal with inheritance and that it wasn't trivial to convert some credentials providers since they didn't expose all the necessary attributes.